### ISSUE TITLE
Disposable dependencies registered as InstancePerLifetimeScope are not disposed after a request

### ISSUE THREAD

**Issue by robbmaclean** [autofac/Autofac.AspNetCore.Multitenant#3]
First, I am a big fan of Autofac , and your multitenant extension is brilliant.  Thank you!

I have converted a large (to me) application over to a multitenant service registration design and I couldn't be happier, except that it appears that my `IDisposable` resources registered as "InstancePerLifetimeScope" are no longer being disposed by Autofac after each request.

I have created 2 very basic asp dotnet core projects to demonstrate, using the Visual Studio 2017 .NET Core Web Application (Web API) template: 
- Project A: Demonstrates Autofac *without* multitenancy disposing my `IDisposable` class after each request.
- Project B: Demonstrates the issue I am humbly reporting - Autofac with multitenancy *not* disposing my `IDisposable` class.  Perhaps this is by design, or perhaps I using the multitenant container and middleware incorrectly, but I was expecting the same behavior with the `SomethingDisposable` registration after adding multitenancy.

Each project has the following structure, and the code pasted below for each project is from `Program.cs`.

![image](https://user-images.githubusercontent.com/25302868/42598165-3e039c2a-8510-11e8-9448-19fb00985784.png)

The `ValuesController` does nothing noteworthy except take my `IDisposable` class `SomethingDisposable` in the constructor.

**Project A**

```cs
using System;
using Autofac.Extensions.DependencyInjection;
using Autofac;
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace AutofacDisposal
{
    /// <summary>
    /// Program
    /// </summary>
    public class Program
    {
        public static void Main(string[] args)
        {
            BuildWebHost(args).Run();
        }

        public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseKestrel()
                .UseStartup<Startup>()
                .Build();
    }

    /// <summary>
    /// Startup
    /// </summary>
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            services.AddMvc();
            var builder = new ContainerBuilder();
            builder.Populate(services);
            builder.RegisterType<SomethingDisposable>()
                .AsSelf()
                .InstancePerLifetimeScope();
            var container = builder.Build();
            return new AutofacServiceProvider(container);
        }

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            app.UseMvc();
        }
    }

    /// <summary>
    /// A class that implements IDisposable and is registered as InstancePerLifetimeScope
    /// </summary>
    public class SomethingDisposable : IDisposable
    {
        private readonly ILogger<SomethingDisposable> _logger;

        public SomethingDisposable(ILoggerFactory loggerFactory)
        {
            _logger = loggerFactory.CreateLogger<SomethingDisposable>();
            _logger.LogInformation("Initializing SomethingDisposable");
        }

        public void Dispose()
        {
            _logger.LogInformation("Disposing SomethingDisposable");
        }
    }
}

```

The log shows my `IDisposable` class being disposed:

![image](https://user-images.githubusercontent.com/25302868/42597624-372ebd14-850e-11e8-91b6-df5a8dcc02c7.png)

**Project B**

This project is what I believe demonstrations the problem, if this is not user error.  I don't actually register services for my fake tenant 123, for brevity.  

```cs
using System;
using Autofac.Extensions.DependencyInjection;
using Autofac;
using Autofac.Multitenant;
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace AutofacDisposalMultitenant
{
    /// <summary>
    /// Program
    /// </summary>
    public class Program
    {
        public static void Main(string[] args)
        {
            BuildWebHost(args).Run();
        }

        public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseKestrel()
                .UseAutofacMultitenantRequestServices(() => Startup.MultitenantContainer)
                .UseStartup<Startup>()
                .Build();
    }

    /// <summary>
    /// Startup
    /// </summary>
    public class Startup
    {
        public static MultitenantContainer MultitenantContainer { get; set; }

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            services.AddMvc();
            var builder = new ContainerBuilder();
            builder.Populate(services);
            builder.RegisterType<SomethingDisposable>()
                .AsSelf()
                .InstancePerLifetimeScope();
            var container = builder.Build();
            MultitenantContainer = new MultitenantContainer(new MyTenantIdentificationStrategy(), container);
            return new AutofacServiceProvider(MultitenantContainer);
        }

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            app.UseMvc();
        }
    }

    /// <summary>
    /// My mock ITenantIdentificationStrategy
    /// </summary>
    public class MyTenantIdentificationStrategy : ITenantIdentificationStrategy
    {
        public bool TryIdentifyTenant(out object tenantId)
        {
            tenantId = 123;
            return true;
        }
    }

    /// <summary>
    /// A class that implements IDisposable and is registered as InstancePerLifetimeScope
    /// </summary>
    public class SomethingDisposable : IDisposable
    {
        private readonly ILogger<SomethingDisposable> _logger;

        public SomethingDisposable(ILoggerFactory loggerFactory)
        {
            _logger = loggerFactory.CreateLogger<SomethingDisposable>();
            _logger.LogInformation("Initializing SomethingDisposable");
        }

        public void Dispose()
        {
            _logger.LogInformation("Disposing SomethingDisposable");
        }
    }
}
```

The log shows my `IDisposable` class being created, but not disposed:

![image](https://user-images.githubusercontent.com/25302868/42597736-9f268dde-850e-11e8-9ac5-d6228eedb07d.png)

Here is the csproj file for Project B:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Autofac" Version="4.8.1" />
    <PackageReference Include="Autofac.AspNetCore.Multitenant" Version="1.0.1" />
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="4.2.2" />
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.8" />
  </ItemGroup>

  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.4" />
  </ItemGroup>

</Project>
```

Thank you, kindly!

**Comment by tillig** [comment]
I just updated the `samples/Sandbox` app to have the code (basically) from your specified replication.

- A dependency that's disposable, registered as `InstancePerLifetimeScope`, and logs its disposal.
- A controller that uses the dependency.
- Two tenants that are configured to use the dependency.

**Clone this repo, fire up the Sandbox app, and try it out three ways**:

- Anonymous / no tenant: http://localhost:35115/api/values
- Tenant A (has a specific version of the dependency configured): http://localhost:35115/api/values?tenant=a
- Tenant C (falls back to default): http://localhost:35115/api/values?tenant=c

In all cases I can see the tenant being properly identified and the dependency is getting disposed.

![Sandbox logging shows disposal](https://user-images.githubusercontent.com/1156571/42603364-6b42b48e-8522-11e8-9ffa-ebeeaef027c4.PNG)

The differences in the Sandbox, from what I can tell, are:

- Sandbox is `netcoreapp2.1` using `Microsoft.AspNetCore.All` 2.1.2 (which you should update to for security reasons).
- Sandbox has an older version of `Microsoft.VisualStudio.Web.CodeGeneration.Tools`. That shouldn't matter since that never gets called.
- Sandbox dependency implements an interface. I tried deleting the interface from the project and registering only `AsSelf` but it didn't matter; the dependency was always disposed.
- Sandbox doesn't _always_ identify the tenant. If there's no request, there's no tenant identified. I tried hardcoding this to always identify a tenant and it didn't make a difference; the dependency was always disposed.
- Sandbox is built using .NET Core SDK 2.1.300. I don't know what version you're using. It shouldn't matter.

Basically, I can't repro this given your code. The only pieces of your code I don't have are:

- Your controller.
- Your tenant-specific registrations.

Given that, I have to believe there's something going on in there.

Here's what I'll recommend:

1. Try out the Sandbox app to verify you see what I see.
2. See if you can modify the Sandbox app to cause the failure you're seeing. If you can, submit a PR so I can see the differences. (It'd be to illustrate the change you made, not as something I'd want to pull into the project.)
3. Set up a real, minimal, working repro in GitHub somewhere and paste a link here. I'll need to see a whole project where this is happening.

Sorry I don't have an answer, I just am not seeing what you're seeing. Did I repro it wrong?

**Comment by robbmaclean** [comment]
Thank you for the quick and detailed reply!  I cloned the repo and verified that disposal is working properly, as you left it in `develop` branch.  

Per your recommendation, I upgraded my project to target `netcoreapp2.1`, and package `Microsoft.AspNetCore.All` to verison `2.1.2`.  I was pleasantly surprised that this simple change seems to have solved the problem for me - I wish I could provide more info about why. 

I was able to reproduce the problem in the "Sandbox" project by downgrading TargetFramework to `netcoreapp2.0` and downgrading package `Microsoft.AspNetCore.All` to verison `2.0.8`.  

I don't seem to have permission to push a branch in order to create a pull request to demonstrate, but the only file that I changed to reproduce the issue is `Sandbox.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Compile Remove="wwwroot\**" />
    <Content Remove="wwwroot\**" />
    <EmbeddedResource Remove="wwwroot\**" />
    <None Remove="wwwroot\**" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Autofac" Version="4.8.1" />
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="4.2.2" />
    <PackageReference Include="Autofac.Multitenant" Version="4.2.0" />
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.8" />
  </ItemGroup>
  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="1.0.1" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\..\src\Autofac.Integration.AspNetCore.Multitenant\Autofac.Integration.AspNetCore.Multitenant.csproj" />
  </ItemGroup>

</Project>
```

Thank you!

**Comment by tillig** [comment]
Interesting. I do see it, too, when using metapackage 2.0.8. Really odd. Guessing they fixed something in there, though I'm not sure what. If someone wants to dig in to ASP.NET Core and figure it out, awesome. Maybe I'll do it, maybe I won't.

Anyway... I guess if someone else runs into this, the answer is "upgrade."

**Comment by tillig** [comment]
Oh, wait, I may be able to manually dispose things. Let me see if I can figure that out.

**Comment by tillig** [comment]
I figured it out.

Super long and complex discussion aside, basically the internals of ASP.NET Core have been changing around how and when they handle disposal of the request lifetime scope. I wasn't handling that right for all versions. I think I figured something out that should work across all the versions and I'll get that pushed shortly.

What a deal. 😩

**Comment by tillig** [comment]
[Pushed v1.0.2.](https://www.nuget.org/packages/Autofac.AspNetCore.Multitenant/1.0.2)
